main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala

```diff
@@ -0,0 +1,716 @@
+    case KindedAst.Spec(_, _, _, _, fparams, _, tpe, eff, tconstrs, econstrs, loc) =>
+
+      val InfResult(infConstrs, infTpe, infEff, infRenv) = infResult
+
+      // The initial substitution maps from formal parameters to their types
+      val initialSubst = fparams.foldLeft(Substitution.empty) {
+        case (acc, KindedAst.FormalParam(sym, mod, tpe, src, loc)) => acc ++ Substitution.singleton(sym.tvar.sym, openOuterSchema(tpe)(Level.Top, flix))
+      }
+
+      // Wildcard tparams are not counted in the tparams, so we need to traverse the types to get them.
+      val allTparams = tpe.typeVars ++ eff.typeVars ++ fparams.flatMap(_.tpe.typeVars) ++ econstrs.flatMap(_.tpe2.typeVars)
+
+      // The rigidity environment is made up of:
+      // 1. rigid variables from the context (e.g. from instance type parameters)
+      // 2. rigid variables from type inference (e.g. regions)
+      // 3. rigid variables from declared function type parameters
+      val renv = allTparams.foldLeft(infRenv ++ renv0) {
+        case (acc, Type.Var(sym, _)) => acc.markRigid(sym)
+      }
+
+      // The class and equality environments are made up of:
+      // 1. constraints from the context (e.g. constraints on instances and classes, plus global constraints)
+      // 2. constraints from the function signature
+      val cenv = expandClassEnv(cenv0, tconstrs ++ tconstrs0)
+      val eqEnv = expandEqualityEnv(eqEnv0, econstrs) // TODO ASSOC-TYPES allow econstrs on instances
+
+      // We add extra constraints for the declared type and effect
+      val tpeConstr = TypingConstraint.Equality(tpe, infTpe, Provenance.ExpectType(expected = tpe, actual = infTpe, loc))
+      val effConstr = TypingConstraint.Equality(eff, infEff, Provenance.ExpectEffect(expected = eff, actual = infEff, loc))
+      val constrs = tpeConstr :: effConstr :: infConstrs
+
+
+      resolve(constrs, initialSubst, renv, cenv, eqEnv).flatMap {
+        case ResolutionResult(subst, deferred, _) =>
+          Debug.stopLogging()
+
+          // If there are any constraints we could not resolve, then we report an error.
+          // TODO ASSOC-TYPES here we only consider the first error
+          deferred match {
+            case Nil => Result.Ok(subst)
+            case TypingConstraint.Equality(tpe1, tpe2, prov) :: _ => Err(toTypeError(UnificationError.MismatchedTypes(tpe1, tpe2), prov))
+            case TypingConstraint.Class(sym, tpe, loc) :: _ => Err(TypeError.MissingInstance(sym, tpe, renv, loc))
+            case TypingConstraint.Purification(_, _, _, _, _, _) :: _ => throw InternalCompilerException("unexpected purificaiton error", loc)
+          }
+      }.toValidation
+  }
```

Comment:
Can we reorder the functions so the file can be read top to bottom?

Here is ok with the helper functions expandClassEnv and a few others, but then I think the main resolve function should come.